File: java-to-js/src/Enums.java
import java.util.List;
import java.util.Set;

public void main() {
  System.out.println(STR."Roles: \{List.of(Role.values())} Permissions: \{List.of(Permission.values())}");
  System.out.println(STR."Admin Permissions: \{Role.ADMIN.permissions}");

  System.out.println(Role.ADMIN.hasPermission(Permission.DELETE));
  System.out.println(Role.VIEWER.hasPermission(Permission.READ));
  System.out.println(Role.EDITOR.hasPermission(Permission.DELETE));
}

enum Role {
  ADMIN(Permission.READ, Permission.WRITE, Permission.DELETE),
  EDITOR(Permission.READ, Permission.WRITE),
  VIEWER(Permission.READ);

  final Set<Permission> permissions;

  Role(Permission... permissions) {
    this.permissions = Set.of(permissions);
  }

  boolean hasPermission(Permission permission) {
    return permissions.contains(permission);
  }
}

enum Permission {
  READ,
  WRITE,
  DELETE
}

File: java-to-js/src/enums.ts
enum Role {
  ADMIN = "ADMIN",
  EDITOR = "EDITOR",
  VIEWER = "VIEWER",
}

enum Permission {
  READ = "READ",
  WRITE = "WRITE",
  DELETE = "DELETE",
}

const rolePermissions: Record<Role, Permission[]> = {
  [Role.ADMIN]: [Permission.READ, Permission.WRITE, Permission.DELETE],
  [Role.EDITOR]: [Permission.READ, Permission.WRITE],
  [Role.VIEWER]: [Permission.READ],
};

function hasPermission(role: Role, permission: Permission): boolean {
  return rolePermissions[role].includes(permission);
}

console.log("Admin Permissions", rolePermissions[Role.ADMIN]);

console.log(hasPermission(Role.ADMIN, Permission.DELETE));
console.log(hasPermission(Role.VIEWER, Permission.READ));
console.log(hasPermission(Role.EDITOR, Permission.DELETE));
